// @sanity/runtime-cli/dist/commands/blueprints/doctor.js
import { Flags } from '@oclif/core';
import { baseFlags, RuntimeCommand, unhide } from '../../baseCommands.js';
import config from '../../config.js';
import { blueprintDoctorCore } from '../../cores/blueprints/doctor.js';
export default class DoctorCommand extends RuntimeCommand {
    static description = 'Diagnose potential issues with Blueprint configuration';
    static examples = [];
    static flags = {
        path: unhide(baseFlags.path),
        fix: Flags.boolean({
            description: 'Interactively fix configuration issues',
            default: false,
        }),
        json: unhide(baseFlags.json),
        verbose: unhide(baseFlags.verbose),
    };
    async run() {
        const { token } = config;
        const result = await blueprintDoctorCore({
            bin: this.config.bin,
            log: (message) => this.log(message),
            token,
            flags: this.flags,
        });
        const { success, error } = result;
        if (!success)
            this.error(error);
        return result.data;
    }
}
